src/elastio-snap.c

```diff
@@ -5186,6 +5185,98 @@ static int __handle_bdev_mount_writable(const char __user *dir_name, const struc
 	return ret;
 }
 
+// The function below is called to gain or restore the ownership over a block device and resolves
+// the following bug related to the dormant snapshot functionality:
+//
+// [ Issue https://github.com/elastio/elastio-snap/issues/126 ]
+//
+//  - at the first mount, the block device is owned by the parent driver
+//  - after the snapshot creation, we update ownership to elastio driver
+//  - because of this, v5.13+ kernels did additional module_put() for our module at umount
+//
+//  To resolve this, when active and umount, we switch the ownership back to the parent, hence
+//  making the umount independent from the perspective of the elastio driver. When the device
+//  is mounted again, re-gain ownership over it.
+//
+//  Apart from it, we do this to prevent the kernel from freeing our block_device_operations
+//  installed (f.e, if the device is removed from the system)
+
+#define OWNERSHIP_TO_PARENT 0
+#define OWNERSHIP_TO_DRIVER 1
+
+static int bdev_switch_ownership(const char __user *dir_name, int follow_flags, int ownership)
+{
+	int i, ret = 0;
+	int lookup_flags = 0;
+	struct path path = {};
+	struct snap_device *dev;
+	struct block_device *bdev;
+	char bdev_name[BDEVNAME_SIZE];
+
+	if(!(follow_flags & UMOUNT_NOFOLLOW)) lookup_flags |= LOOKUP_FOLLOW;
+
+	ret = user_path_at(AT_FDCWD, dir_name, lookup_flags, &path);
+	if(ret){
+		ret = -EINVAL;
+		LOG_DEBUG("error finding path");
+		goto out;
+	}else if(path.dentry != path.mnt->mnt_root){
+		ret = -ENODEV;
+		LOG_DEBUG("path specified isn't a mount point");
```

Comment:
Why not log this interesting path?

Reply:
1. This will be excessive, as we already have a path right above this message:
```
[13489.845105] elastio-snap: detected block device umount: /run/credentials/mnt.mount : 10
[13489.845116] elastio-snap: error finding path
[13489.845119] elastio-snap: umount returned: -2
[13489.845295] elastio-snap: detected block device umount: /run/user/1000/credentials/mnt.mount : 10
[13489.845306] elastio-snap: error finding path
[13489.845309] elastio-snap: umount returned: -2
[13489.845652] elastio-snap: post umount check succeeded
[13489.845655] elastio-snap: umount returned: 0
```
2. To do this simple thing as a derivation of mount path, we'll need to add `copy_from_user()` + additional code checks, which also seems unreasonable for such a small improvement (it's not used anywhere). I'm not sure if it makes sense to do this.

Reply:
:+1: